feat: view similar photos

normana10 opened the pull request:
## Description

My use-case may be a bit odd, but I expect this may be quite useful for others as well

In my collection I have _about_ 30k duplicates (mostly re-downloads scraped from social media like Facebook)

When I come across one of these low-res images that I _know_ I have a higher-resolution version of somewhere, I wanted to be able to use the same logic the de-duping utility does to view similar images (not to mention the fact that the scraped photos have the date of the scrape and not when the photo was taken)

And to boot this lookup is actually _quite_ faster than the "Search by context" because the embedding for the example asset _already exists_!

## How Has This Been Tested?

- Open any asset
- Click the "more" kebab
- Click the "View similar photos" menu item
- Observe that you are now seeing similar photos in descending similarity

(I'll note that of course the "example" image should always come _first_ as it'll have the highest similarity score... but I'm going to say that's a feature rather than a bug)

<details><summary><h2>Screenshots</h2></summary>

<img width="1916" height="965" alt="image" src="https://github.com/user-attachments/assets/24d4bda2-59eb-4583-8f07-c31c6391cc8f" />
<img width="1918" height="968" alt="image" src="https://github.com/user-attachments/assets/22e6289e-9498-4e6f-b84e-612c2c4c1437" />

</details>

## API Changes
The `/search/smart` endpoint now takes an `example_asset_id` field that is used to retrieve similar assets

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation if applicable
- [x] I have no unrelated changes in the PR.
- [x] I have confirmed that any new dependencies are strictly necessary.
- [ ] I have written tests for new code (if applicable)
- [x] I have followed naming conventions/patterns in the surrounding code
- [x] All code in `src/services/` uses repositories implementations for database calls, filesystem operations, etc.
- [x] All code in `src/repositories/` is pretty basic/simple and does not have any immich specific logic (that belongs in `src/services/`)

normana10 committed:
Enable filteing by example

normana10 committed:
Drop `@GenerateSql` for `getEmbedding`?

normana10 committed:
Improve error message

bo0tzz commented:
How does this PR relate to #17061?

jrasm91 commented on server/src/repositories/search.repository.ts:
```diff
@@ -293,6 +293,12 @@ export class SearchRepository {
   }
 
+  async getEmbedding(assetId: string) {
+    return this.db.transaction().execute(async (trx) => {
+      return trx.selectFrom('smart_search').selectAll().where('assetId', '=', assetId).executeTakeFirst();
+    });
+  }
+
```
this can just be `return this.db.selectFrom('smart_search' ...)` and you can add a `@GenerateSql({ params: [DummyValue.UUID] })` thing as well.

normana10 replied:
Got it, no `set local ...`, no need for the transaction

And added the `@GenerateSql`

jrasm91 commented on server/src/services/search.service.ts:
```diff
@@ -113,14 +113,25 @@ export class SearchService extends BaseService {
+      }
+    } else if (dto.exampleAssetId) {
+      const assetEmbedding = (await this.searchRepository.getEmbedding(dto.exampleAssetId))?.embedding;
+      if (assetEmbedding === undefined) {
```
```suggestion
      if (!assetEmbedding) {
```

jrasm91 commented on server/src/services/search.service.ts:
```diff
@@ -113,14 +113,25 @@ export class SearchService extends BaseService {
-      });
-      this.embeddingCache.set(key, embedding);
+    let embedding;
+    if (dto.query) {
```
This validation should move to the dto. You can throw an error if one of the two fields are not provided there. Then here, you can just do something like:
```ts
let embedding;
if(dto.exampleAssetId) {
  embedding = ...
}
  
  
if(dto.query) {
  embedding = ...
}
 ```

jrasm91 commented on server/src/dtos/search.dto.ts:
```diff
@@ -198,8 +198,12 @@ export class StatisticsSearchDto extends BaseSearchDto {
+
+  @IsString()
+  @Optional()
+  exampleAssetId?: string;
```
Maybe `queryAssetId` would be better?

normana10 commented:
@bo0tzz 

That... is a very good point

Sorry I hadn't noticed that PR

I suppose they both aim to achieve similar goals but use _slightly_ different methods

#17061 appears to intend to use structured language to parse the example assetId out of the `query` parameter where I have added a separate field for it to live in (that also seems to be kind of exactly what you suggested in https://github.com/immich-app/immich/pull/17061#issuecomment-2748001645 which is entertaining because I was completely unaware of it until right now 😅)

That said, I'm interested in your thoughts on what I've got here!

I've only PR'd this part, but my _intent_ is to use it to expand the capability of the dedupe util (sneak peek screenshot below, ignore all the extra fields I added to the individual dupe assets, those are probably only really helpful to _me_)
<img width="1457" height="1171" alt="image" src="https://github.com/user-attachments/assets/2ae63147-6aba-461f-b61c-cd6856e1816c" />

normana10 committed:
PR Feedback

normana10 committed:
Sort en.json

jrasm91 commented on server/src/services/search.service.ts:
```diff
@@ -113,14 +113,25 @@ export class SearchService extends BaseService {
+        });
+        this.embeddingCache.set(key, embedding);
+      }
+    } else if (dto.queryAssetId) {
```
You need to validate this, probably with requireAccess(Permission.AssetRead) or whatever.

normana10 replied:
Is that not already being handled by the `@Authenticated({ permission: Permission.AssetRead })` up in [the controller layer function](https://github.com/immich-app/immich/blob/main/server/src/controllers/search.controller.ts#L58) that calls this function?

Or should I say, to have gotten this far, have we not already validated the caller has `Permission.AssetRead`?

jrasm91 replied:
They do, yes. That validates the API key or session token has that permission, but it does not validate or imply that queryAssetId is an asset they should have access to. There is a difference between being able to read your assets vs all/any assets in the system.

normana10 replied:
Ooooh neat didn't even know that mechanism existed

Implemented 👍

normana10 committed:
Add SQL

normana10 committed:
Fix lint

normana10 committed:
Drop test that is no longer valid

normana10 committed:
Fix i18n file sorting

normana10 committed:
Fix TS error

normana10 committed:
Add a `requireAccess` before pulling the embedding

jrasm91 reviewed (commented):
Looks good to me. Just a few small things with the decorators, which I am happy to just fix and merge myself if you want

normana10 committed:
Fix decorators

normana10 committed:
Merge remote-tracking branch 'upstream' into enable-filter-to-similar-assets

normana10 committed:
Run `make open-api`

alextran1502 committed:
Merge branch 'main' into enable-filter-to-similar-assets
